File: julia/p004.jl
# Largest palindrome product

DIGITS = 3


function solve()
    return maximum((string(p) == reverse(string(p))) ? p : 0
                   for p in (x * y for x in 10^(DIGITS - 1):(10^DIGITS - 1)
                   for y in x:(10^DIGITS - 1)))
end


print(Int(solve()))

File: haskell/p004.hs
-- Largest palindrome product

products :: [Int]
products = do
    [ x * y | x <- [1..limit], y <- [1..limit]]
    where
        limit = 10^(3 :: Int) - 1

palindromeTest :: Int -> Bool
palindromeTest x = reverse (show x) == show x

palindromes :: [Int]
palindromes = filter palindromeTest products


main :: IO ()
main = do
    print (maximum palindromes)
